### PR TITLE
chore: custom ESLint rule to disable sorting for `ref` attribute to keep it as last property in JSX to ensure it is up-to-date

### PR DESCRIPTION
**Pull request by Elijbet**
**Related Issue:** #7659

## Summary
Custom ESLint rule to provide a more granular prop ordering mechanism by forcing the `ref` attribute to be kept as the last property so we can ensure it is up-to-date.

Added a test file for the new custom rules: the file references different use cases, which are 1 good and 1 bad usage files.

### CODE REVIEW

**Comment by Elijbet** [comment]
> It doesn't look like jsx-eslint/eslint-plugin-react has kept their docs up to date. There is a prop called reservedList available for the eslint react/jsx-sort-props that should allow "ref" to be at the bottom of a multiline list of attributes.
> 
> ```json
> "react/jsx-sort-props": [
>       "error",
>       {
>         reservedList: ["ref"],
>       },
>  ],
> ```

I would rather not use an undocumented prop, it doesn't feel like a good practice.
It's possible they just forgot to document it, but it is also possible that it wasn’t documented intentionally and only exists in the source code because they don’t want it being used for whatever reason. Have we not ever done the latter before?

But I'll defer to @driskull and @jcfranco for perspective.

**Comment by alisonailea** [comment]
I ran quick test on my local branch before I made the comment. However, after opening a clean branch off of Main and running some additional tests, I don’t think it’s working the way I thought.

That said, I think it would be better to open an issue and PR to eslint-jsx/eslint-react with a "reservedLast" prop (or something) rather than making (and then having to maintain) a new eslint plugin.

**Comment by driskull** [comment]
> I would rather not use an undocumented prop, it doesn't feel like a good practice.

Agreed.

> I ran quick test on my local branch before I made the comment. However, after opening a clean branch off of Main and running some additional tests, I don’t think it’s working the way I thought.

`reservedList` sounds like its something that isnt allowed to be used?

> That said, I think it would be better to open an issue and PR to eslint-jsx/eslint-react with a "reservedLast" prop (or something) rather than making (and then having to maintain) a new eslint plugin.

Agreed. We should open an issue/PR but in the meantime we could do our own thing with a custom plugin. Who knows if they would add or accept the feature we are requesting.

File: packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts
Comment (driskull) [inline packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts:7]: Maybe:

` "Ensures the ref attribute is ordered last in a JSXElement to keep it up-to-date",`?

File: packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts
Comment (driskull) [inline packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts:38]: remove console log.

File: packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts
Comment (driskull) [inline packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts:16]: remove comments

File: packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts
Comment (driskull) [inline packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts:43]: `The "ref" attribute should be the last attribute in a JSXElement.`

**Review by driskull — COMMENTED** [review]
Looks good! Maybe the wording could be cleaned up a little.

**Review by alisonailea — APPROVED** [review]

File: packages/calcite-components/.eslintrc.cjs
Comment (jcfranco) [inline packages/calcite-components/.eslintrc.cjs:15]: Why is `.eslintrc.cjs` ignored here?

File: packages/eslint-plugin-calcite-components/README.md
Comment (jcfranco) [inline packages/eslint-plugin-calcite-components/README.md:38]: `shell` might be more appropriate since this is not JS.

File: packages/eslint-plugin-calcite-components/README.md
Comment (jcfranco) [inline packages/eslint-plugin-calcite-components/README.md:54]: Can you add info from https://github.com/Esri/calcite-design-system/pull/6530 to provide more context? The rule description and its doc file could also benefit from this extra info.

File: packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts
Comment (jcfranco) [inline packages/eslint-plugin-calcite-components/src/rules/enforce-ref-last-prop.ts:34]: Could you add more info from https://github.com/Esri/calcite-design-system/pull/7584 to the error message?

File: packages/eslint-plugin-calcite-components/tests/lib/rules/enforce-ref-last-prop/enforce-ref-last-prop.good.tsx
Comment (jcfranco) [inline packages/eslint-plugin-calcite-components/tests/lib/rules/enforce-ref-last-prop/enforce-ref-last-prop.good.tsx:8]: Suggestion: could you also add `tabIndex` to tie in the [scenario](https://github.com/jcfranco/stencil-ref-called-in-prop-order-instead-of-after-node-being-initialized/blob/main/src/components/my-component/my-component.tsx#L12-L15) that this rule helps avoid?

**Review by jcfranco — APPROVED** [review]
🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️
🕶️😎😎😎😎🕶️😎🕶️🕶️🕶️🕶️🕶️😎🕶️😎😎😎😎🕶️🕶️😎😎😎🕶️🕶️😎😎🕶️🕶️😎🕶️🕶️🕶️😎🕶️😎😎😎😎🕶️😎🕶️
🕶️😎🕶️🕶️😎🕶️😎🕶️🕶️🕶️🕶️🕶️😎🕶️😎🕶️🕶️🕶️🕶️😎🕶️🕶️🕶️🕶️😎🕶️🕶️😎🕶️😎😎🕶️😎😎🕶️😎🕶️🕶️🕶️🕶️😎🕶️
🕶️😎😎😎😎🕶️😎🕶️🕶️😎🕶️🕶️😎🕶️😎😎😎🕶️🕶️🕶️😎😎🕶️🕶️😎🕶️🕶️😎🕶️😎🕶️😎🕶️😎🕶️😎😎😎🕶️🕶️😎🕶️
🕶️😎🕶️🕶️😎🕶️😎🕶️😎🕶️😎🕶️😎🕶️😎🕶️🕶️🕶️🕶️🕶️🕶️🕶️😎🕶️😎🕶️🕶️😎🕶️😎🕶️🕶️🕶️😎🕶️😎🕶️🕶️🕶️🕶️🕶️🕶️
🕶️😎🕶️🕶️😎🕶️🕶️😎🕶️🕶️🕶️😎🕶️🕶️😎😎😎😎🕶️😎😎😎🕶️🕶️🕶️😎😎🕶️🕶️😎🕶️🕶️🕶️😎🕶️😎😎😎😎🕶️😎🕶️
🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️🕶️

File: packages/eslint-plugin-calcite-components/tests/lib/rules/enforce-ref-last-prop/enforce-ref-last-prop.good.tsx
Comment (Elijbet) [reply packages/eslint-plugin-calcite-components/tests/lib/rules/enforce-ref-last-prop/enforce-ref-last-prop.good.tsx:8]: That's a good point! ☉

File: packages/eslint-plugin-calcite-components/tests/lib/rules/enforce-ref-last-prop/enforce-ref-last-prop.wrong.tsx
Comment (jcfranco) [inline packages/eslint-plugin-calcite-components/tests/lib/rules/enforce-ref-last-prop/enforce-ref-last-prop.wrong.tsx:15]: `tabIndex` should go here to match the good case.

File: packages/eslint-plugin-calcite-components/README.md
Comment (jcfranco) [inline packages/eslint-plugin-calcite-components/README.md:54]: The `this updates ref usage...` part doesn't really apply here because the rule isn't changing the code. I think it'll read better if we flip it a bit where we describe what the rule does `This rule ensures...` and then follow it with
 
`This updates ref usage to workaround is a workaround for a Stencil bug (<add link to bug*>) where ref is...`

**Review by jcfranco — APPROVED** [review]
This should be good to merge once the last comments are addressed. 🚀

**Pull request merged**
